libraries/chain/include/eosio/chain/pending_snapshot.hpp

```diff
@@ -33,9 +33,16 @@ class pending_snapshot {
       return snapshots_dir / fc::format_string(".incomplete-snapshot-${id}.bin", fc::mutable_variant_object()("id", block_id));
    }
 
-   T finalize(const chain::controller& chain) const {
+   T finalize(const block_id_type& lib_id, const chain::controller& chain) const {
+      auto lib_num = chain::block_header::num_from_id(lib_id);
+      auto block_num = chain::block_header::num_from_id(block_id);
+
+      EOS_ASSERT(lib_num >= block_num, chain::snapshot_finalization_exception,
+                 "finalize called for non-irreversible block ${bn}:${bid}", ("bn", block_num)("bid", block_id));
+
+      // finalize called before forkdb is pruned of non-irreversible blocks, so this can find a non-irreversible block
       auto block_ptr = chain.fetch_block_by_id(block_id);
-      auto in_chain = (bool) block_ptr;
+      auto in_chain = !!block_ptr && (lib_num > block_num || lib_id == block_id);
```

Comment:
I'm not immediately understanding the rationale for this check and the assert above. What conditions mean that a simple `bool in_chain = lib_id == block_id` isn't sufficient? I guess tihs is what your second paragraph is getting at.

Reply:
My thoughts exactly. The more I think about this, I think I should simplify this instead of leaving this "dead" code.

Reply:
I wonder if there might be some corner case where where you start from a snapshot in HEAD mode and request a snapshot before receiving any new blocks. Maybe you'll never get on on_reverisble for that initial block since it was already irreversible?

Reply:
Hmm, that does seem possible. I think we could just EOS_ASSERT on that with a message that LIB not signaled. You should have a snapshot for that HEAD since that is what you started with to begin with.  Might be possible to miss it as well by switching from IRREVERSIBLE to HEAD and the producer plugin on active when the blocks are applied. Again an EOS_ASSERT to the caller that LIB was missed seems like an ok fall-back for those cases.

Reply:
In that case maybe instead of the `EOS_ASSERT` here, the logic that iterates through the pending snapshots could be modified,
https://github.com/AntelopeIO/spring/blob/17b468887e2db1301bf3892ba1caa72e21f2ebc1/libraries/chain/snapshot_scheduler.cpp#L52
It's the `<=` that causes this weirdness.

So in this loop we can then check if it's `<` (only will occur for these startup quirks afaict so far) then we can skip calling `finialize()` altogether (maybe still log an error :woman_shrugging:), and then if it's `==` the logic becomes a simple `bool in_chain = lib_id == block_id` I think?

Just an idea.

Reply:
oh we can't do that since it's finalize that removes the pending snapshot. womp womp nevermind

Reply:
For the cases outlined, I guess best to just leave it as is.